python/pyarrow/tests/test_array.py

```diff
@@ -688,6 +688,69 @@ def test_dictionary_from_numpy():
             assert d2[i].as_py() == dictionary[indices[i]]
 
 
+def test_dictionary_to_numpy():
+    expected = pa.array(
+        ["foo", "bar", None, "foo"]
+    ).to_numpy(zero_copy_only=False)
+    a = pa.DictionaryArray.from_arrays(
+        pa.array([0, 1, None, 0]),
+        pa.array(['foo', 'bar'])
+    )
+    assert (a.to_numpy(zero_copy_only=False) == expected).all()
+
+    with pytest.raises(pa.ArrowInvalid):
+        # to_numpy takes for granted that when zero_copy_only=True
+        # there will be no nulls.
+        # If that changes, nulls handling will have to be updated in to_numpy
+        # as we won't be able to rely anymore on decoding the DictionaryArray
+        # to handle nulls.
+        a.to_numpy(zero_copy_only=True)
+
+    anonulls = pa.DictionaryArray.from_arrays(
+        pa.array([0, 1, 1, 0]),
+        pa.array(['foo', 'bar'])
+    )
+    expected = pa.array(
+        ["foo", "bar", "bar", "foo"]
+    ).to_numpy(zero_copy_only=False)
```

Comment:
```suggestion
    expected = np.array(["foo", "bar", "bar", "foo"], dtype=object)
```

I think we can just define the expected here? (it's also a bit simpler in code)

(and the same for the others)

Reply:
My goal was mostly to test that "wathever decoding a normal `Array` leads to, decoding a `DictionaryArray` leads to the same exact result". I felt it was more future proof than testing for the actual value. But if you feel it's better to test for the actual resulting numpy array I'll gladly change that.

Reply:
OK, either way is fine for me